src/styles/style-effector.ts

```diff
@@ -0,0 +1,191 @@
+ * limitations under the License.
+ */
+
+import {ASTWalker, ExpressionNode, FunctionNode} from './parsers';
+
+interface AnyObserver {
+  observe(): void;
+  disconnect(): void;
+}
+
+const $instances = Symbol('instances');
+const $activateListener = Symbol('activateListener');
+const $deactivateListener = Symbol('deactivateListener');
+const $notifyInstances = Symbol('notifyInstances');
+const $notify = Symbol('notify');
+const $scrollCallback = Symbol('callback');
+
+type ScrollObserverCallback = () => void;
+
+/**
+ * This internal helper is intended to work as a reference-counting manager of
+ * scroll event listeners. Only one scroll listener is ever registered for all
+ * instances of the class, and when the last ScrollObserver "disconnects", that
+ * event listener is removed. This spares us from thrashing
+ * the {add,remove}EventListener API (the binding cost of these methods has been
+ * known to show up in performance anlyses) as well as potential memory leaks.
+ */
+class ScrollObserver {
+  private static[$notifyInstances]() {
+    for (const instance of ScrollObserver[$instances]) {
+      instance[$notify]();
+    }
+  }
+  private static[$instances]: Set<ScrollObserver> = new Set();
+  private static[$activateListener]() {
+    window.addEventListener('scroll', this[$notifyInstances], {passive: true});
+  }
+  private static[$deactivateListener]() {
+    window.removeEventListener('scroll', this[$notifyInstances]);
+  }
+
+  private[$scrollCallback]: ScrollObserverCallback;
+
+  constructor(callback: ScrollObserverCallback) {
+    this[$scrollCallback] = callback;
+  }
+
+  /**
+   * Listen for scroll events. The configured callback (passed to the
+   * constructor) will be invoked for subsequent global scroll events.
+   */
+  observe() {
+    if (ScrollObserver[$instances].size === 0) {
+      ScrollObserver[$activateListener]();
+    }
+    ScrollObserver[$instances].add(this);
+  }
+
+  /**
+   * Stop listening for scroll events.
+   */
+  disconnect() {
+    ScrollObserver[$instances].delete(this);
+    if (ScrollObserver[$instances].size === 0) {
+      ScrollObserver[$deactivateListener]();
+    }
+  }
+
+  private[$notify]() {
+    this[$scrollCallback]();
+  };
+}
+
+
+export type EnvironmentState = 'window-scroll';
+
+export type StyleEffectorCallback = (record: EnvironmentChangeRecord) => void;
+export interface EnvironmentChangeRecord {
+  relatedState: EnvironmentState;
+}
+
+type EnvironmentDependencies = {
+  [key in EnvironmentState]?: AnyObserver
+};
+
+const $computeStyleCallback = Symbol('computeStyleCallback');
+const $astWalker = Symbol('astWalker');
+const $dependencies = Symbol('dependencies');
+
+const $scrollHandler = Symbol('scrollHandler');
+const $onScroll = Symbol('onScroll');
+
+/**
+ * The StyleEffector is configured with a callback that will be invoked at the
+ * optimal time that some array of CSS expression ASTs ought to be evaluated.
```

Comment:
Speaking of avoiding contractions, what is AST?

Reply:
["abstract syntax tree"](https://en.wikipedia.org/wiki/Abstract_syntax_tree)

For us CS/engineer types, it is a very common acronym (and "abstract syntax tree" is a bit of a mouthful).